src/coreclr/jit/lsrariscv64.cpp

```diff
@@ -286,10 +286,65 @@ int LinearScan::BuildNode(GenTree* tree)
         case GT_MOD:
         case GT_UMOD:
         case GT_DIV:
-        case GT_MULHI:
         case GT_UDIV:
         {
             srcCount = BuildBinaryUses(tree->AsOp());
+
+            GenTree* divisorOp = tree->gtGetOp2();
+            if (!varTypeIsFloating(tree->TypeGet()) &&
+                !(divisorOp->IsIntegralConst(0) || divisorOp->GetRegNum() == REG_R0))
+            {
+                bool needTemp = false;
+                if (divisorOp->isContainedIntOrIImmed())
+                {
+                    needTemp = true; // divisorReg
+                }
+                else if (tree->OperIsCommutative())
+                {
+                    if (tree->gtGetOp1()->isContainedIntOrIImmed())
+                        needTemp = true; // reg1
+                }
+
+                if (tree->gtOper == GT_DIV || tree->gtOper == GT_MOD)
+                {
+                    bool checkDividend = true;
+                    // Do we have an immediate for the 'divisorOp'?
+                    if (divisorOp->IsCnsIntOrI())
+                    {
+                        ssize_t intConstValue = divisorOp->AsIntCon()->gtIconVal;
+                        if (intConstValue != -1)
```

Comment:
nit: I have troubles with understanding this part. Maybe adding one extra comment explaining meaning of -1 and impact on next check could help a bit.

Reply:
To know whether we need a temp it's enough to mimic the logic from CodeGen::genCodeForDivMod (which also doesn't explain the meaning of -1:) There is some explanation in the condition under which ArithmeticException flag is set, which is GenTree::CanDivOrModPossiblyOverflow.

But basic idea IMO is clear: integer division can overflow only for div by -1, any other other divisor results in a smaller number. And it does overflow if dividend is MinValue because -MinValue > MaxValue.